File: algorithms/medium/minimum_time_to_collect_all_apples_in_a_tree.cpp
class Solution {
public:
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
      unordered_map<int, vector<int>> mp;
      for (auto& edge : edges) {
        mp[edge[0]].push_back(edge[1]);
        mp[edge[1]].push_back(edge[0]);
      }
      return max(aux(mp, 0, hasApple, -1) - 2, 0);
    }

  int aux(unordered_map<int, vector<int>>& mp, int curr, vector<bool>& hasApple, int parent) {
    int ct = 0;
    for (auto child : mp[curr]) {
      if (child == parent) continue;
      ct += aux(mp, child, hasApple, curr);
    }
    if (ct > 0) return ct + 2;
    if (hasApple[curr]) return 2;
    return 0;
  }
};

File: algorithms/medium/minimum_time_to_collect_all_apples_in_a_tree.ex
defmodule Solution do
  @spec min_time(n :: integer, edges :: [[integer]], has_apple :: [boolean]) :: integer
  def min_time(n, edges, has_apple) do
    has_apple =
      has_apple
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {b, i}, acc -> Map.put(acc, i, b) end)

    edges
    |> Enum.reduce(%{}, fn [a, b], acc ->
      acc
      |> Map.update(a, [b], &[b | &1])
      |> Map.update(b, [a], &[a | &1])
    end)
    |> aux(0, has_apple, -1)
    |> :erlang.-(2)
    |> max(0)
  end

  def aux(tree, curr, has_apple, parent) do
    Map.get(tree, curr, [])
    |> Enum.reduce(0, fn child, acc ->
      acc + if child == parent, do: 0, else: aux(tree, child, has_apple, curr)
    end)
    |> then(fn ct ->
      cond do
        ct > 0 -> ct + 2
        has_apple[curr] -> 2
        true -> 0
      end
    end)
  end
end
